src/py/flwr/client/secure_aggregation/secaggplus_handler.py

```diff
@@ -0,0 +1,326 @@
+
+from flwr.client.client import Client
+from flwr.client.numpy_client import NumPyClient
+from flwr.common import (
+    bytes_to_ndarray,
+    ndarray_to_bytes,
+    ndarrays_to_parameters,
+    parameters_to_ndarrays,
+)
+from flwr.common.logger import log
+from flwr.common.secure_aggregation.crypto.shamir import create_shares
+from flwr.common.secure_aggregation.crypto.symmetric_encryption import (
+    bytes_to_private_key,
+    bytes_to_public_key,
+    decrypt,
+    encrypt,
+    generate_key_pairs,
+    generate_shared_key,
+    private_key_to_bytes,
+    public_key_to_bytes,
+)
+from flwr.common.secure_aggregation.quantization import quantize
+from flwr.common.secure_aggregation.secaggplus_utils import (
+    pseudo_rand_gen,
+    share_keys_plaintext_concat,
+    share_keys_plaintext_separate,
+)
+from flwr.common.secure_aggregation.weights_arithmetic import (
+    factor_weights_combine,
+    weights_addition,
+    weights_mod,
+    weights_multiply,
+    weights_subtraction,
+)
+from flwr.common.typing import FitIns, Value
+
+from .handler import SecureAggregationHandler
+
+_stages = ["setup", "share keys", "collect masked input", "unmasking"]
+
+
+@dataclass
+# pylint: disable=invalid-name
+# pylint: disable-next=too-many-instance-attributes
+class _State:
+    sample_num: int = 0
+    sid: int = 0
+    share_num: int = 0
+    threshold: int = 0
+    test_drop: bool = False
+    clipping_range: float = 0.0
+    target_range: int = 0
+    mod_range: int = 0
+    b_share_dict: Dict[int, bytes] = field(default_factory=dict)
+    sk1_share_dict: Dict[int, bytes] = field(default_factory=dict)
+    shared_key_2_dict: Dict[int, bytes] = field(default_factory=dict)
+    sk1: bytes = b""
+    pk1: bytes = b""
+    sk2: bytes = b""
+    pk2: bytes = b""
+    b: bytes = b""
+    public_keys_dict: Dict[int, Tuple[bytes, bytes]] = field(default_factory=dict)
+    client: Optional[Union[Client, NumPyClient]] = None
+
+
+# pylint: enable=invalid-name
+
+
+class SecAggPlusHandler(SecureAggregationHandler):
+    """Message handler for the SecAgg+ protocol."""
+
+    _shared_state = _State()
+    _current_stage = "unmasking"
+
+    def handle_secure_aggregation(
+        self, named_values: Dict[str, Value]
+    ) -> Dict[str, Value]:
+        """Handle incoming message and return results, following the SecAgg+ protocol.
+
+        Parameters
+        ----------
+        named_values : Dict[str, Value]
+            The named values retrieved from the SecureAggregation sub-message
+            of Task message in the server's TaskIns.
+
+        Returns
+        -------
+        Dict[str, Value]
+            The final/intermediate results of the SecAgg+ protocol.
+        """
+        if not isinstance(self, (Client, NumPyClient)):
+            raise TypeError(
+                "The subclass of SecAggPlusHandler must be "
+                "the subclass of Client or NumPyClient."
+            )
+        stage = str(named_values.pop("stage"))
```

Comment:
We can certainly do that. But I am still a bit confused about it. If the message does not contain the expected values, should we throw an exception saying that? Because SA will not work and no privacy will be leaked if incoming messages are not as expected.

Reply:
The least we should do is stop the execution as early as possible if the required fields are not provided. This should be communicated back to the server as an error response.